src/operator/tensor/matrix_op.cc

```diff
@@ -763,7 +763,7 @@ parameter values:
 .add_arguments(ClipParam::__FIELDS__());
 
 NNVM_REGISTER_OP(_backward_clip)
-.set_num_inputs(1)
+.set_num_inputs(2)
```

Comment:
I'm surprised to see this was not caught previously. Is there any unit test to unvail this bug?

Reply:
I noticed it when trying to attach monitor to the executor which had clip - it failed with assertion about number of inputs mismatch.
I am not aware of any other test that could uncover this - I don't think there are any checks for this.

Reply:
I think the bug in this case is the inconsistency between set_num_inputs(1) for the _backward_clip node and the definition of the forward node:
```
NNVM_REGISTER_OP(clip)
...
 .set_num_inputs(1)
 .set_num_outputs(1)
...
.set_attr<nnvm::FGradient>("FGradient", ElemwiseGradUseIn{ "_backward_clip" })
```
Rather than a unittest, I think the framework should detect this problem in the general case.  The following CHECK_EQ added to ./src/operator/operator_common.h flags this problem in test_clip(), but I haven't determined whether the entire CI is compatible with it:
```
inline std::vector<nnvm::NodeEntry>CreateNodeEntries(
  nnvm::NodePtr pNode,
  const std::vector<nnvm::NodeEntry>* pOgrads = nullptr,
  const std::vector<nnvm::NodeEntry>* pInputs = nullptr) {
  if (pOgrads)
    pNode->inputs.insert(pNode->inputs.end(), pOgrads->begin(), pOgrads->end());

  if (pInputs)
    pNode->inputs.insert(pNode->inputs.end(), pInputs->begin(), pInputs->end());

  CHECK_EQ(pNode->inputs.size(), pNode->num_inputs()) <<
           "Mismatch in number of created vs. expected node inputs.";

  std::vector<nnvm::NodeEntry> ret;
  for (uint32_t i = 0; i < pNode->num_outputs(); ++i)
    ret.emplace_back(nnvm::NodeEntry{pNode, i, 0});

  return ret;
}
```

Reply:
That is a good idea to have a check like this, but probably in a subsequent PR.

Reply:
I have submitted the separate PR https://github.com/apache/incubator-mxnet/pull/15834 that includes this check of actual vs. expected node inputs.  Initial finding is that many nodes are similarly non-compliant, like clip(), so best as a separate PR that doesn't slow down this one.